### PR TITLE
refactor: unify APK building and installing process

### PR DESCRIPTION
**Pull request by krizzu**
Summary:
---------

Hey!

While working on #1038, I've noticed that there are two different ways to build and install APK: one with `--deviceId` provided and one without. This PR unifies that process.

### What has changed

- Moved all methods related to APK build and install to `buildAndInstallApk.ts` (including utility functions). 
- Removed old and unused code. 
- Exported two new functions : 

    - `buildApk` - builds APK using provided variants.
    - `installApk` - checks if APK has been build for specified variant. If it is, use `adb` to install. Otherwise use gradle's `install*` task to build and install.

To Do:
----------

- [x] cleanup
- [x] tests

Test Plan:
----------

Green CI.

### CODE REVIEW

**Comment by thymikee** [comment]
How's it going, do you need any help here?

**Comment by Krizzu** [comment]
Hey, working on Jest tests and manual testing too - should be here soonish

File: packages/platform-android/src/commands/runAndroid/index.ts
Comment (grabbou) [inline packages/platform-android/src/commands/runAndroid/index.ts]: I am wondering if we could somehow handle #124 here?

Comment (grabbou) [reply packages/platform-android/src/commands/runAndroid/index.ts]: Also, these tests might be useful https://github.com/react-native-community/cli/pull/934/files

Comment (Krizzu) [reply packages/platform-android/src/commands/runAndroid/index.ts]: Is #124 still an issue?

**Comment by Krizzu** [comment]
This one is ready for some reviews, thanks!

File: packages/platform-android/src/commands/runAndroid/__tests__/buildAndInstallApk.test.ts
Comment (thymikee) [inline packages/platform-android/src/commands/runAndroid/__tests__/buildAndInstallApk.test.ts]: How about using `writeFiles` and `cleanup` helpers from our e2e dir?

Comment (thymikee) [reply packages/platform-android/src/commands/runAndroid/__tests__/buildAndInstallApk.test.ts]: nvm, I see you mock the fs, that's fine

File: packages/platform-android/src/commands/runAndroid/buildAndInstallApk.ts
Comment (thymikee) [inline packages/platform-android/src/commands/runAndroid/buildAndInstallApk.ts]:
```suggestion
import {logger, CLIError} from '@react-native-community/cli-tools';
```

File: packages/platform-android/src/commands/runAndroid/index.ts
Comment (dcalhoun) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: @Krizzu wanted to follow up on your question and state that #124 does appear to be occurring for me. I have [captured my details on the issue](https://github.com/react-native-community/cli/issues/124#issuecomment-611133214).

Comment (Krizzu) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: @dcalhoun 
I'm addressing your issue in this PR, but to make it correct, we'd have to put some constraints/assumptions first:
- No flavour mixing (multiple flavour)
- flavour name must be lowercase

Otherwise, we'd have to find a better way (either different CLI arguments or something else) than what [I did here](https://github.com/react-native-community/cli/commit/9100b764a132ff17f663af83fa0d01c00141eb97#diff-a1c62b130fddb803e053cba3ec05f980R189) 

cc @thymikee

Comment (dcalhoun) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: @Krizzu thanks! The second constraint should be easy enough to meet. The first a bit more difficult, as we build our Android app for two different stores (Google Play, Amazon), different environments (Production, Staging), and toggle debuggability with build variants. I suppose there may be a way to duplicate config, rather than composing it, but I'll have to look into it some.

Thank you regardless of the response and efforts.

Comment (Krizzu) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: @dcalhoun the constraints I mention are by no means set in stone. Actually, if we only keep the second one, we could actually support flavour mixing.

Comment (dcalhoun) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: Ah, OK. The casing of build variant names most definitely carries almost no value to me. If enforcing lowercase allows for composing build variants, that would be my preferences.

Comment (Krizzu) [reply packages/platform-android/src/commands/runAndroid/index.ts:268]: I've added flavour mixing support, FYI @dcalhoun @thymikee

**Comment by Krizzu** [comment]
@thymikee @Esemesek 
Any idea why [those `doctor` e2e tests](https://dev.azure.com/react-native-community/cli/_build/results?buildId=1193&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=5caf77c8-9b10-50ef-b5c7-ca89c63e1c86&l=493) fails?

**Comment by thymikee** [comment]
Azure fails with "EPERM: operation not permitted, mkdir 'D:\'" when we try to create a temporary directory for testing, maybe something changed there? Sounds like Azure changing something, would be good to verify this in other PRs

**Comment by Esemesek** [comment]
I think this test fails randomly. We consider Windows failures as non-blocking since most of them are just flakiness in the Azure environment.

**Comment by grabbou** [comment]
Do you feel like giving this one a go @thymikee @Esemesek before shipping? Since it's ready and failures are non-related, I'd love to have a final approval from your side.

**Comment by grabbou** [comment]
Since this code is ready to go and this is our second attempt and rewriting this relatively major part of `run-android`, I am going to take it over and do the last round of testing myself.

I am hoping we can ship it next week, it should resolve some really annoying issues.

**Comment by driiftkiing** [comment]
@thymikee: i have rebased the branch of this pr with the master into this branch https://github.com/tokyodrift1993/react-native-community-cli/tree/feature/buildApkFixes

**Comment by thymikee** [comment]
@driiftkiing can you send a PR amending this one? :)

**Comment by driiftkiing** [comment]
> @driiftkiing can you send a PR amending this one? :)

just creating a new one? or something else?

**Comment by Krizzu** [comment]
@driiftkiing @thymikee 
Been a while 🤖 
I just rebased the code, tests are passing. I'm yet to test this, making sure that it works properly.

**Comment by Krizzu** [comment]
Ran some tests on custom android project, trying different configuration of build variants (with and without flavors) and it works. 
Would love to have someone else do the sanity check as well 🙏

**Comment by grabbou** [comment]
Thanks @Krizzu, I am going to do another pass at it this week and ship it!

**Comment by driiftkiing** [comment]
I will do also some tests, if i got the time

**Review by andriytsaryov — APPROVED** [review]

**Comment by thymikee** [comment]
Thanks for the review @andriytsaryov. @grabbou I guess we can finally go ahead and merge it

**Comment by grabbou** [comment]
Yes, that's on my todo list for tomorrow. When ready, I will create a new release.

**Review by grabbou — APPROVED** [review]

File: packages/platform-android/src/commands/runAndroid/runOnAllDevices.ts
Comment (grabbou) [inline packages/platform-android/src/commands/runAndroid/runOnAllDevices.ts:49]: I believe this code is already moved out to `buildApk` and `installApk` respectively, e.g. `args.port` and `createInstallError`. I am going to go ahead and remove it.

**Comment by pas256** [comment]
I'd love to be able to run

```
npx react-native run-android --variant=developmentDebug --deviceId 931X1X9JW
```

and have it work. This PR seems to solve that. Anything holding this up getting merged in?

**Comment by grabbou** [comment]
No, @pas256, this PR will be pulled in soon, once we merge a PR removing `link` and `unlink`.

**Comment by driiftkiing** [comment]
if anyone is interested to use this PR until it is merged:
- created patch: https://gist.github.com/driiftkiing/09f3bd7bb125e6436a5caba7834f3a99
- source: https://github.com/tokyodrift1993/react-native-community-cli/tree/krizzu/buildApkFixes--merged-v6.3.0
- merged [v6.3.0](https://github.com/react-native-community/cli/tree/v6.3.0)
- (fixed lint & tests)
- should be compatible with [react-native@0.67.2](https://github.com/facebook/react-native/tree/v0.67.2)

\
command to run:
```shell
yarn react-native run-android --variant=developmentDebug
```

**Comment by TMisiukiewicz** [comment]
#1739 closes this one

**Comment by krizzu** [comment]
🫡

**Pull request closed without merging**
